jwql/website/apps/jwql/oauth.py

```diff
@@ -0,0 +1,177 @@
+        user : dict
+            A dictionary of user credentials
+
+        Returns
+        -------
+        fn : function
+            The decorated function
+        """
+
+        # If user is currently anonymous, require a login
+        if user["anon"]:
+            # Redirect to oauth login
+            redirect_uri = os.path.join(get_base_url(), 'authorize')
+            return JWQL_OAUTH.mast_auth.authorize_redirect(request, redirect_uri)
+
+        return fn(request, user)
+
+    return check_auth
+
+
+def auth_info(fn):
+    """A decorator function that will return user credientials along
+    with what is returned by the original function.
+
+    Parameters
+    ----------
+    fn : function
+        The function to decorate
+
+    Returns
+    -------
+    user_info : function
+        The decorated function
+    """
+
+    def user_info(request):
+        """Store authenticated user credentials in a cookie and return
+        it.  If the user is not authenticated, store no credentials in
+        the cookie.
+
+        Parameters
+        ----------
+        request : HttpRequest object
+            Incoming request from the webpage
+
+        Returns
+        -------
+        fn : function
+            The decorated function
+        """
+
+        cookie = request.COOKIES.get("ASB-AUTH")
+
+        # If user is authenticated, return user credentials
+        if cookie is not None:
+            response = requests.get(
+                'https://{}/info'.format(get_config()['auth_mast']),
+                headers={'Accept': 'application/json',
+                         'Authorization': 'token {}'.format(cookie)})
+            response = response.json()
+
+        # If user is not authenticated, return no credentials
+        else:
+            response = {'ezid' : None, "anon": True}
+
+        return fn(request, response)
+
+    return user_info
+
+
+def register_oauth():
+    """Register the ``jwql`` application with the ``auth.mast``
+    authentication service.
+
+    Returns
+    -------
+    oauth : Object
+        An object containing methods to authenticate a user, provided
+        by the ``auth.mast`` service.
+    """
+
+    # Get configuration parameters
+    client_id = get_config()['client_id']
+    client_secret = get_config()['client_secret']
+    auth_mast = get_config()['auth_mast']
+
+    # Register with auth.mast
+    oauth = OAuth()
+    client_kwargs = {'scope' : 'mast:user:info'}
+    oauth.register(
+        'mast_auth',
+        client_id='{}'.format(client_id),
+        client_secret='{}'.format(client_secret),
+        access_token_url='https://{}/oauth/access_token?client_secret={}'.format(auth_mast, client_secret),
+        access_token_params=None,
+        refresh_token_url=None,
+        authorize_url='https://{}/oauth/authorize'.format(auth_mast),
+        api_base_url='https://{}/1.1/'.format(auth_mast),
+        client_kwargs=client_kwargs)
+
+    return oauth
+
+
+JWQL_OAUTH = register_oauth()
```

Comment:
Is there a particular reason this global variable is at the bottom, rather than the top?

Reply:
@laurenmarietta Yes, since it is calling the `register_oauth()` function, the `JWQL_OAUTH` variable must be defined after that function is defined.  Putting it at the top would cause a `NameError` stating that `register_oauth` is not defined.

One might argue this is bad practice.  I guess its a matter of convenience.  Doing it this way allows a user to do:

```python
from jwql.website.apps.jwql.oauth import JWQL_OAUTH
```

instead of having to do:

```python
from jwql.website.apps.jwql.oauth import register_oauth
jwql_oauth = register_oauth()
```

I'm not exactly sure the best way to handle this 🤷‍♂️

Reply:
Ahh, I see now. I've gotten so used to functions (where the order doesn't matter in the same way) that I completely forgot that it would matter in this case!

This seems fine to me then, even if it's a little funny.